hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java

```diff
@@ -852,6 +869,59 @@ private void initializeDagger(
         }
     }
 
+    /**
+     * Given the {@link BlockStreamInfo} context from a {@link State}, infers the block hash of the
+     * last block that was incorporated in this state.
+     * @param state the state to use
+     * @return the inferred block hash
+     */
+    private Bytes startBlockHashFrom(@NonNull final State state) {
+        final var blockStreamInfo = state.getReadableStates(BlockStreamService.NAME)
+                .<BlockStreamInfo>getSingleton(BLOCK_STREAM_INFO_KEY)
+                .get();
+        requireNonNull(blockStreamInfo);
+        // Three of the four ingredients in the block hash are directly in the BlockStreamInfo; that is,
+        // the previous block hash, the input tree root hash, and the start of block state hash
+        final var prevBlockHash = blockHashByBlockNumber(
+                blockStreamInfo.trailingBlockHashes(),
+                blockStreamInfo.blockNumber() - 1,
+                blockStreamInfo.blockNumber() - 1);
+        requireNonNull(prevBlockHash);
+        final var leftParent = combine(prevBlockHash, blockStreamInfo.inputTreeRootHash());
+        // The fourth ingredient, the output tree root hash, is not directly in the BlockStreamInfo, but
+        // we can recompute it based on the tree hash information and the fact the last output item in
+        // the block was devoted to putting the BlockStreamInfo itself into the state
+        final var outputTreeRootHash = outputTreeRootHashFrom(blockStreamInfo);
+        final var rightParent = combine(outputTreeRootHash, blockStreamInfo.startOfBlockStateHash());
+        return combine(leftParent, rightParent);
+    }
+
+    /**
+     * Given a {@link BlockStreamInfo} context, computes the output tree root hash that must have been
+     * computed at the end of the block that the context describes, assuming the final output block item
+     * was the state change that put the context into the state.
```

Comment:
Are there any conditions under which that assumption might not hold -- or will it always definitely be the last output item ?

Reply:
No exceptions...the `BlockStreamInfo` singleton update will always be the last block output item---see [here](https://github.com/hashgraph/hedera-services/blob/15431-restart-block-hash-mgmt/hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/BlockStreamManagerImpl.java#L224).

Reply:
Awesome! Thanks @tinker-michaelj  😎

Reply:
Agreed @thomas-swirlds-labs , @jsync-swirlds also suggested this---the [header comments for this message](https://github.com/hashgraph/hedera-services/blob/15431-restart-block-hash-mgmt/hapi/hedera-protobufs/services/state/blockstream/block_stream_info.proto#L52) now state it must come last:
```
 * The state change to update this singleton MUST be the last "output"
 * block item in this block.
```